From a4dd1b470744ac5bba7621994f5a46ac5e5800c4 Mon Sep 17 00:00:00 2001
From: Alexey Kodanev <aleksei.kodanev@bell-sw.com>
Date: Tue, 16 Aug 2022 16:31:08 +0300
Subject: [PATCH] abuild.conf: enable stack-protector-strong and
 _FORTIFY_SOURCE=2

Add hardening checks using annobin/annocheck. The !annobin option
can be set to disable the test and building with the plugin.
---
 abuild.in                      | 80 +++++++++++++++++++++++++++++++++-
 default.conf                   |  4 +-
 tests/testrepo/dbgpkg/APKBUILD |  7 ++-
 3 files changed, 86 insertions(+), 5 deletions(-)

diff --git a/abuild.in b/abuild.in
index 793c5c5..a8b1841 100644
--- a/abuild.in
+++ b/abuild.in
@@ -2071,7 +2071,9 @@ default_dbg() {
 		if ! [ -e "$pkgbasedir/.dbg-tmp/$ino" ]; then
 			local tmp=$pkgbasedir/.dbg-tmp/${src##*/}
 			${CROSS_COMPILE}objcopy --only-keep-debug "$src" "$dst"
-			${CROSS_COMPILE}objcopy --add-gnu-debuglink="$dst" --strip-unneeded -R .comment "$src" "$tmp"
+			${CROSS_COMPILE}objcopy --add-gnu-debuglink="$dst" --strip-unneeded \
+				-R .comment -R .gnu.build.attributes* \
+				"$src" "$tmp"
 			# preserve attributes, links
 			cat "$tmp" > "$src"
 			rm "$tmp"
@@ -2295,6 +2297,7 @@ rootpkg() {
 	do_fakeroot "$abuild_path" $forceroot $color_opt $keep_build $verbose \
 		set_source_date \
 		package \
+		hardening_check \
 		prepare_subpackages \
 		prepare_language_packs \
 		prepare_package \
@@ -2712,13 +2715,66 @@ stripbin() {
 
 		[ "$osabi" != "STANDALONE" ] || continue
 		local XATTR=$(getfattr --match="" --dump "${filename}")
-		"${stripcmd}" "${filename}"
+		"${stripcmd}" --remove-section=.gnu.build.attributes* "${filename}"
 		if [ -n "$XATTR" ]; then
 			echo "$XATTR" | "$SETFATTR" --restore=-
 		fi
 	done
 }
 
+hardening_check()
+{
+	if options_has "!annobin" || [ "${subpkgarch:-$pkgarch}" = "noarch" ]; then
+		return 0
+	fi
+
+	cd "${subpkgdir:-$pkgdir}" || return 1
+
+	cd "$pkgdir"
+
+	local annocheck_skip_tr=
+	if [ -n "$annocheck_skip" ]; then
+		local s
+		for s in $annocheck_skip; do
+			annocheck_skip_tr="$annocheck_skip_tr $(echo $s | tr -d [:space:])"
+		done
+	fi
+
+	if ! options_has "!fortify" && [ "$CLIBC" = "musl" ] && \
+		[ ! -f "/usr/include/fortify/string.h" ]; then
+		warning "Hardened: missing fortify-headers"
+	fi
+
+	scanelf --recursive --nobanner --etype "ET_DYN,ET_EXEC" . | \
+		while read type path; do
+
+		fname="$(basename $path)"
+		suffix="${fname##*.}"
+
+		[ "$suffix" ] && [ "$suffix" = "debug" -o "$suffix" = "debuginfo" ] && \
+			continue
+
+		if [ -n "$annocheck_skip_tr" ]; then
+			skipped=0
+
+			for s in $annocheck_skip_tr; do
+				if [ "$s" = "$fname" ]; then
+					skipped=1
+					break
+				fi
+			done
+
+			if [ "$skipped" -eq 1 ]; then
+				msg "Hardened: $path skipped"
+				continue
+			fi
+		fi
+		annocheck --ignore-unknown --ignore-links $path || \
+			warning "Hardened: check failed: $path"
+	done
+	cd - >/dev/null
+}
+
 # simply list target apks
 listpkg() {
 	local name
@@ -3032,6 +3088,26 @@ if [ -n "$DEBUG" ] || subpackage_types_has "dbg"; then
 	options="$options !strip"
 fi
 
+remove_cflags()
+{
+	local f="$@"
+	export CFLAGS="${CFLAGS/$f/}"
+	export CXXFLAGS="${CXXFLAGS/$f/}"
+	export CPPFLAGS="${CPPFLAGS/$f/}"
+}
+
+if options_has "!annobin" || [ -n "$ABUILD_DISABLE_ANNOBIN" ] ; then
+	remove_cflags "-fplugin=annobin"
+fi
+
+if options_has "!stackprot"; then
+	remove_cflags "-fstack-protector-strong"
+fi
+
+if options_has "!fortify"; then
+	remove_cflags "-D_FORTIFY_SOURCE=2"
+fi
+
 if [ -n "$subpkgname" ]; then
 	# If we are handling a sub package then reset subpackages and install
 	origsubpackages="$subpackages"
diff --git a/default.conf b/default.conf
index 04c2540..1945f77 100644
--- a/default.conf
+++ b/default.conf
@@ -1,5 +1,5 @@
-export CFLAGS="-Os -fstack-clash-protection -Wformat -Werror=format-security"
-export CXXFLAGS="-Os -fstack-clash-protection -Wformat -Werror=format-security -D_GLIBCXX_ASSERTIONS=1"
+export CFLAGS="-fplugin=annobin -Os -fstack-clash-protection -Wformat -Werror=format-security -fstack-protector-strong -D_FORTIFY_SOURCE=2"
+export CXXFLAGS="-fplugin=annobin -Os -fstack-clash-protection -Wformat -Werror=format-security -D_GLIBCXX_ASSERTIONS=1 -fstack-protector-strong -D_FORTIFY_SOURCE=2"
 export LDFLAGS="-Wl,--as-needed,-O1,--sort-common"
 export GOFLAGS="-buildmode=pie -modcacherw -trimpath -buildvcs=false"
 # Do note that these should work with at least GDC and LDC
diff --git a/tests/testrepo/dbgpkg/APKBUILD b/tests/testrepo/dbgpkg/APKBUILD
index e090c35..e3c8b7b 100644
--- a/tests/testrepo/dbgpkg/APKBUILD
+++ b/tests/testrepo/dbgpkg/APKBUILD
@@ -18,7 +18,12 @@ prepare() {
 }
 
 build() {
-	${CC:-gcc} -o hello hello.c
+	${CC:-gcc} \
+		-Os \
+		-fplugin=annobin \
+		-fstack-protector-strong \
+		-D_FORTIFY_SOURCE=2 \
+		-o hello hello.c
 }
 
 check() {
-- 
2.41.0

